streambed/src/crypto.rs

```diff
@@ -1,35 +1,41 @@
 //! Helpers to encrypt and decrypt things, particularly message payloads.
 
 use aes::{
-    cipher::{generic_array::GenericArray, BlockEncrypt, KeyInit},
+    cipher::{generic_array::GenericArray, BlockEncrypt},
     Aes128,
 };
 use rand::RngCore;
+use sha2::{Digest, Sha256, Sha512};
+
+use hmac::Hmac;
+type HmacSha256 = Hmac<Sha256>;
 
 const BLOCKSIZE: usize = 16;
 
+pub const KEY_SIZE: usize = 16;
 pub const SALT_SIZE: usize = 12;
 
 /// Generate a random salt for the required SALT_SIZE.
-pub fn salt<T>(rng: &mut T) -> Vec<u8>
+pub fn salt<T>(rng: &mut T) -> [u8; SALT_SIZE]
 where
     T: RngCore,
 {
-    let mut bytes = vec![0; SALT_SIZE];
+    let mut bytes = [0; SALT_SIZE];
     rng.fill_bytes(&mut bytes);
     bytes
 }
 
 /// Performs an AES-128 ECB decryption using a salt and a 16 byte key
 #[inline]
-pub fn decrypt(bytes: &mut [u8], key: &[u8], salt: &[u8]) {
+pub fn decrypt(bytes: &mut [u8], key: &[u8; KEY_SIZE], salt: &[u8; SALT_SIZE]) {
     encrypt(bytes, key, salt)
 }
 
 /// Performs an AES-128 ECB encryption using a salt and a 16 byte key
-pub fn encrypt(bytes: &mut [u8], key: &[u8], salt: &[u8]) {
+pub fn encrypt(bytes: &mut [u8], key: &[u8; KEY_SIZE], salt: &[u8; SALT_SIZE]) {
     debug_assert!(salt.len() == SALT_SIZE);
```

Comment:
Does this also need to run in a release build to prevent inadvertent running with a smaller salt?

Reply:
Don’t understand. The salt size is fixed intentionally.

Reply:
I think that the salt comes from config but you're only checking that the length is correct in a debug build so it could be configured incorrectly in a release situation and this assert wouldn't fire. Maybe I'm wrong though.

Reply:
Ah, the size is checked at compile time. Is the debug_assert useful then?

Reply:
Ah, forgot to take out that assert. Thanks.